hoomd/mpcd/SlitPoreGeometry.h

```diff
@@ -0,0 +1,237 @@
+ */
+class __attribute__((visibility("default"))) SlitPoreGeometry
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param H Channel half-width
+         * \param L Pore half-length
+         * \param bc Boundary condition at the wall (slip or no-slip)
+         */
+        HOSTDEVICE SlitPoreGeometry(Scalar H, Scalar L, boundary bc)
+            : m_H(H), m_L(L), m_bc(bc)
+            { }
+
+        //! Detect collision between the particle and the boundary
+        /*!
+         * \param pos Proposed particle position
+         * \param vel Proposed particle velocity
+         * \param dt Integration time remaining (inout).
+         *
+         * \returns True if a collision occurred, and false otherwise
+         *
+         * \post The particle position \a pos is moved to the point of reflection, the velocity \a vel is updated
+         *       according to the appropriate bounce back rule, and the integration time \a dt is decreased to the
+         *       amount of time remaining.
+         *
+         * The passed value of \a dt must be the time taken to arrive at pos. The returned value of \a dt will be
+         * less than this time.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+            /* First check that the particle ended up inside the pore or walls.
+             * sign.x is +1 if outside pore in +x, -1 if outside pore in -x, and 0 otherwise.
+             * sign.y is +1 if outside walls in +z, -1 if outside walls in -z, and 0 otherwise. */
+            const char2 sign = make_char2((pos.x >= m_L) - (pos.x <= -m_L),
+                                          (pos.z >  m_H) - (pos.z <  -m_H));
+            // exit early if collision didn't happen
+            if (sign.x != 0 || sign.y == 0)
+                {
+                dt = Scalar(0);
+                return false;
+                }
+
+            // times to either hit the pore in x, or the wall in z
+            Scalar2 dts;
+            if (vel.x != Scalar(0))
+                {
+                // based on direction moving, could only have hit one of these edges
+                const Scalar xw = (vel.x > 0) ? -m_L : m_L;
+                dts.x = (pos.x - xw) / vel.x;
+                }
+            else
+                {
+                // no collision
+                dts.x = Scalar(-1);
+                }
+            if (vel.z != Scalar(0))
+                {
+                dts.y = (pos.z - sign.y*m_H) / vel.z;
+                }
+            else
+                {
+                // no collision
+                dts.y = Scalar(-1);
+                }
+
+            // determine if collisions happend with the x or z walls.
+            // if both occur, use the one that happened first (leaves the most time)
+            // this neglects coming through the corner exactly, but that's OK to an approx.
```

Comment:
Sorry, the language in my comment here was pretty vague. For the case where the particle comes through the corner, `hits.x && hits.y` and `dts.x == dts.y`. In this case, the branch taken will be the normal in x because `dts.x` is not smaller than `dts.y`, and so `hits.y` is set to 0. So, the case isn't "neglected" in the sense of ignored... it's "neglected to be treated exactly".

The most accurate thing to do would be to define the normal as the average of the two in this case, but I think the extra branching is not worth it for such a rare case.

Reply:
That's fair.